src/unity/toolkits/coreml_export/neural_net_models_exporter.cpp

```diff
@@ -37,65 +43,275 @@ std::shared_ptr<MLModelWrapper> export_object_detector_model(
+      input_image_pipeline->mutable_type()->mutable_imagetype();
+  input_image_pipeline_feature->set_width(image_width);
+  input_image_pipeline_feature->set_height(image_height);
+  input_image_pipeline_feature->set_colorspace(ImageFeatureType::RGB);
+
+  // Write FeatureDescription for the IOU Threshold input.
+  FeatureDescription* input_iou_pipeline = pipeline_desc->add_input();
+  input_iou_pipeline->set_name("iouThreshold");
+  input_iou_pipeline->set_shortdescription("(optional) IOU Threshold override (default: 0.45)");
+  input_iou_pipeline->mutable_type()->mutable_doubletype();
+
+  // Write FeatureDescription for the Confidence Threshold input.
+  FeatureDescription* input_cthreshold_pipeline = pipeline_desc->add_input();
+  input_cthreshold_pipeline->set_name("confidenceThreshold");
+  input_cthreshold_pipeline->set_shortdescription(
+    "(optional) Confidence Threshold override (default: 0.25)");
+  input_cthreshold_pipeline->mutable_type()->mutable_doubletype();
+
+  // Write FeatureDescription for the Confidence output.
+  FeatureDescription* output_confidence_pipeline = pipeline_desc->add_output();
+  output_confidence_pipeline->set_name("confidence");
+  output_confidence_pipeline->set_shortdescription(
+        "Boxes × Class confidence (see user-defined metadata \"classes\")");
+  ArrayFeatureType* output_confidence_pipeline_feature =
+      output_confidence_pipeline->mutable_type()->mutable_multiarraytype();
+  output_confidence_pipeline_feature->set_datatype(ArrayFeatureType::DOUBLE);
+  auto *confidence_shape_pipeline1 = output_confidence_pipeline_feature->mutable_shaperange()->add_sizeranges();
+  confidence_shape_pipeline1->set_upperbound(-1);
+  auto *confidence_shape_pipeline2 = output_confidence_pipeline_feature->mutable_shaperange()->add_sizeranges();
+  confidence_shape_pipeline2->set_lowerbound(6);
+  confidence_shape_pipeline2->set_upperbound(6);
+
+  // Write FeatureDescription for the Coordinates output.
+  FeatureDescription* output_coordinates_pipeline = pipeline_desc->add_output();
+  output_coordinates_pipeline->set_name("coordinates");
+  output_coordinates_pipeline->set_shortdescription(
+        "Boxes × [x, y, width, height] (relative to image size)");
+  ArrayFeatureType* output_coordinates_pipeline_feature =
+      output_coordinates_pipeline->mutable_type()->mutable_multiarraytype();
+  output_coordinates_pipeline_feature->set_datatype(ArrayFeatureType::DOUBLE);
+  auto *coordinates_shape_pipeline1 = output_coordinates_pipeline_feature->mutable_shaperange()->add_sizeranges();
+  coordinates_shape_pipeline1->set_upperbound(-1);
+  auto *coordinates_shape_pipeline2 = output_coordinates_pipeline_feature->mutable_shaperange()->add_sizeranges();
+  coordinates_shape_pipeline2->set_lowerbound(4);
+  coordinates_shape_pipeline2->set_upperbound(4);
+  
+  // Add NeuralNetwork model to pipeline
+  auto* model1 = model_pipeline.mutable_pipeline()->add_models();
+  *model1 = model_nn;
```

Comment:
This line copies around ~60 MB of data from `model_nn` into the pipeline, which we should make some effort to avoid. For this performance reason, and also for the sake of clarity, I would suggest a little reorganization of this function.

In particular, the protocol buffer APIs are designed to facilitate a top-down approach to filling out a protobuf value. Note that in the original code, we began by allocating the highest level Model instance, and then filling in everything inside using calls to `mutable_*` or `add_*` as appropriate.

Similarly, I suspect this will be simpler if in the NMS case, we follow this outline:
```
CoreML::Specification::Model model_pipeline;
auto* model_nn = model_pipeline->mutable_pipeline->add_models();
// Populate *model_nn.
auto* model_nms = model_pipeline->mutable_pileine->add_models();
// Populate *model_nms
```

I would also really recommend breaking this down into helper functions, to keep things organized. Offhand, one for populating `model_nn` (which can be shared with the non-NMS case), and perhaps some functions for reducing duplication in setting the model inputs and outputs (since Pipeline models explicitly duplicate these).

Reply:
That's a really great point, I've updated the code accordingly. Ty